### PR TITLE
Add package gluon-airtime

### PR DESCRIPTION
**Pull request by corny**
This package adds information about the channel usage to respondd:
- `frequency`: the current frequency
- `active`: time on channel
- `rx`: time spent on receiving
- `tx`: time spent on transmitting
- `busy`: time spent on waiting for transmitting

It works fine ath9k, ath10k does not return the information.

We used this package at our Breminale event Bremen on ~80 nodes (> 2k concurrent clients) to calculate the channel utilization with our [respond-collector](https://github.com/FreifunkBremen/respond-collector/commit/92aac7b7ca55da66e212ce5f71557f6d5d0b2c49#diff-76df96424dfe13846cb6cdcb2a5c7ef5R35).

One strange observation: In some intervals it happens that `Δ(busy) + Δ(rx) + Δ(tx) > Δ(active)`.

### CODE REVIEW

**Comment by corny** [comment]
It might be more useful to calculate the channel usage directly on the node. This should be done in a seperate thread to avoid side effects by requesting the statistics. Otherwise the respond-client has to know the previous values to do the calculation.

**Comment by T-X** [comment]
That sounds great! Have you thought about adding that information to a node's webpage, too? Having a live graph for airtime similar to the RSSI graphs would be awesome and super helpful :-).

**Comment by jplitza** [comment]
This might be more appropriately put into the gluon-packages repository, as there is not really any special ties to gluon (only to respondd, which is also over there).

**Comment by corny** [comment]
New pull request: https://github.com/freifunk-gluon/packages/pull/141

**Comment by corny** [comment]
@T-X That would be nice. I don't think that I have enough time to implement it. Maybe someone else?

**Comment by NeoRaider** [comment]
But the package *is* Gluon-specific! Even when it's only the naming of the interfaces, moving this package to the OpenWrt/LEDE upstream repos doesn't make sense (and the packages repo is for packages that are support to go there at some point).

So the Gluon main repo is the correct place, not the packages repo.

**Comment by corny** [comment]
ok, everything back  :grimacing:

**Comment by jplitza** [comment]
@NeoRaider To me that means that the package has to autodetect the interface names before it can be moved upstream, but its functionality is generic enough that I would see "moving upstream" as its goal, even though things have to be done first. But if you say so…

**Comment by corny** [comment]
I could also use uci to iterate over the wifi-ifaces and find a wifi-iface with 2.4 GHz and another with 5 GHz. Would that help?

**Comment by NeoRaider** [comment]
Using uci is not necessary IMO. Another option to make this independent of Gluon would be to iterate over all radios and add airtime stats for each of them. Instead of "airtime24" and "airtime5", you could have a field "airtime", which is an (unordered) array of stats for all radios. The "frequency" field can then be used to identify the radios.

If you make this independent of Gluon, then indeed this can be added to the *packages* repo. In that case, the package name should not contain Gluon (maybe just call it "respondd-airtime"?).

**Comment by corny** [comment]
I try to get the airtime by radio name. Currently we do:
```c
ifx = if_nametoindex(interface);
NLA_PUT_U32(msg, NL80211_ATTR_IFINDEX, ifx);
```

**Comment by NeoRaider** [comment]
Will this return the cumulative airtime of all interfaces of a single radio, or just the single VIF?

**Comment by corny** [comment]
The return value for all VIFs of a radio are equal. The values come directly from the radio.

**Comment by jplitza** [comment]
I agree that iterating over all radios is preferable to using uci for anything. And while all VIFs for one radio give you the same numbers (which raises the question why this is an `iw dev` command instead of an `iw phy` one), one might have multiple radios for the same frequency band, in which case the frequency isn't a unique indicator for the values. @NeoRaider, you said "unordered array", so you are proposing a JSON structure like this
```json
"airtime": [
  { "frequency": 5180, "active": ...},
  { "frequency": 5200, "active": ...},
  ...
  { "frequency": 5180, "active": ...},
  { "frequency": 5200, "active": ...},
  ...
]
```
where one frequency can show up multiple times (for different radios)?

**Comment by NeoRaider** [comment]
@jplitza, yes, maybe like this.

The current code looks like it only adds a single entry per radio, but the netlink interface should return a full survey of all channels, this confuses me a bit. If we want a full survey, we might also group the entries by radio, and use the frequencies as object keys:
```json
"airtime": [
  {
    "5180": { "active": ...},
    "5200": { "active": ...}
  },
  ...
  {
    "5180": { "active": ...},
    "5200": { "active": ...}
  },
  ...
]
```
or something like this. But I might also misinterpret the code, I didn't have a closer like at it yet.

**Comment by rotanid** [comment]
it might be nice to have it independent from gluon, but the author already said he won't do that work.
so it would be nice to have it in gluon at least, maybe @NeoRaider could take a look on the code from this point of view if no one wants to volunteer to "upstream" this and not only discuss about it....

**Comment by RubenKelevra** [comment]
> One strange observation: In some intervals it happens that Δ(busy) + Δ(rx) + Δ(tx) > Δ(active)

We discovered that too, some years ago. I think it's happening because the times are rounded to a given value, so it might be "0,89" but 1 is recorded. This sums up to high value for each step. But active seems to be recoreded otherwise, like "we're sending now 103 packets at once, counting them all together as one "step".

**Comment by corny** [comment]
@RubenKelevra Thanks a lot. That helps us to understand the data.

**Comment by RubenKelevra** [comment]
@corny that was just my best guess without knowing any details. So prove me wrong if you found anything better.

Else the chip does output the noise-level of all channels it saws. That might be not especially useful with low dataset-times but on the channel which is in use, this might be helpful to discover a wifi which is some channels off or a different type of signal which still reduces the signal noise ratio significantly.

We just screenread all this values pretty dirty from iw which is not especially a pretty solution. ;)

**Comment by jplitza** [comment]
@NeoRaider I just noticed you asked a question in your last comment. Yes, the package only reports airtime numbers for a single frequency, namely the one currently in use. YMMD for different drivers, but ath9k only returns significant numbers for the current frequency:
```
# iw dev client0 survey dump
Survey data from client0
	frequency:			2412 MHz
	channel active time:		26 ms
	channel busy time:		21 ms
	channel receive time:		16 ms
	channel transmit time:		0 ms
Survey data from client0
	frequency:			2417 MHz
Survey data from client0
	frequency:			2422 MHz
[...]
Survey data from client0
	frequency:			2437 MHz [in use]
	noise:				-95 dBm
	channel active time:		1704490481 ms
	channel busy time:		945335227 ms
	channel receive time:		584372518 ms
	channel transmit time:		310956458 ms
Survey data from client0
	frequency:			2442 MHz
Survey data from client0
	frequency:			2447 MHz
[...]
```
Channel 1 was probably used during initialization of the driver/device, and channel 6 is used in our network. No other channel has any numbers.

I still plan on making this iterate over all radios. But in case that takes too long, I just pushed a version that uses the format in my previous comment (not the one in @NeoRaider's previous comment, for the above reasons: We only report one frequency per device per band). Although, to be honest, I don't consider this mergeable as long as the device names are hardcoded, because client interfaces can be (and often are) turned off and thus inexistent.

**Comment by RubenKelevra** [comment]
@jplitza you've holding your wifi wrong. Only 1, 5, 9 and 13 should be used in Germany.

File: package/gluon-airtime/src/airtime.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#include <stdint.h>
+
+struct airtime_time {
+  uint64_t current;
+  uint64_t offset;
+};
+
+struct airtime_result {
+  uint32_t frequency;
+	uint8_t  noise;
```
Comment (rotanid) [inline package/gluon-airtime/src/airtime.h]: tab indention here, 2 spaces everywhere else?

File: package/gluon-airtime/src/airtime.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include <stdint.h>
+
+struct airtime_time {
```
Comment (rotanid) [inline package/gluon-airtime/src/airtime.h]: indention of this struct is two spaces, but everywhere else its tabs?

File: package/gluon-airtime/src/airtime.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  uint64_t current;
```
Comment (rotanid) [inline package/gluon-airtime/src/airtime.h]: indention of this struct is two spaces, but everywhere else its tabs?

File: package/gluon-airtime/src/airtime.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  uint64_t offset;
+};
+
+struct airtime_result {
+  uint32_t frequency;
```
Comment (rotanid) [inline package/gluon-airtime/src/airtime.h]: indention of this struct is two spaces, but everywhere else its tabs?

**Comment by rotanid** [comment]
fyi: tried this package using a WDR3500 a few days ago and it didn't work, neither respondd nor the included demo did output the information. did some quick/short debugging with @corny but no result yet.

**Comment by viisauksena** [comment]
we have something similar in freiburg - using respondd and a shell airtime script which also calculate averages 
i summarize shortly
package -> https://github.com/viisauksena/gluon-airtime
only works for 2G - is heavily redundant (tx,rx,avg..) shows uplink-interface and channel and quality (batman ttvn)

airtime example

**Comment by corny** [comment]
I would like to test the package on multiple devices. Unfortunately I forgot how to build a single package. Does anyone know how to build it if I start with a gluon site repo? Maybe @NeoRaider ?

**Comment by corny** [comment]
Thanks to @Brother-Lal I was able to fix the issues. I tested it succesfully with:
* WR841 v10
* WDR3600
* CPE210

**Comment by jplitza** [comment]
@corny so what issues were there to fix? The only issue remaining, in my opinion, is the hardcoded interface problem, which still persists. And by force pushing, you overwrote my changes. Please stop that.

**Comment by corny** [comment]
The checks in the `survey_airtime_handler()` lead to abortion of the function.

**Comment by jplitza** [comment]
I think this is finished. The interfaces are now autodetected by asking the kernel for a list of wifi interfaces (VIFs, not PHYs) and picking (the first) one for each PHY.

The only question that remains: Should this instead go to the gluon-packages repo, because it doesn't depend on Gluon now in any way?

**Comment by rotanid** [comment]
as already controversially discussed on IRC, we would prefer to also include some average over the last n minutes - this could be used on the status page or in external software like hopglass or meshviewer without having to do much coding in every one of these three examples.
we already implemented that a while ago: https://github.com/Brother-Lal/gluon/commits/airtime

**Comment by corny** [comment]
My opinion: Keep out complexity from the C code and calculate the values you need in your map backend.

**Comment by NeoRaider** [comment]
I also think that the respondd provider should always return a momentary value and not average itself.

This should indeed go into the packages repo (might also make sense to rename it in a way that doesn't contain "gluon", but that doesn't really matter).

**Comment by rotanid** [comment]
i don't think the momentary value and the average contradict each other - our version simply returns both.
well, whatever...

**Comment by Adorfer** [comment]
Everybody can pick what is needed. 
If you do not want to poll every minute, an average could come handy.

**Comment by jplitza** [comment]
I fixed a memory leak that caused havoc in my long running respondd (valgrind doesn't report anything now for the test program) and added a readme detailing the format.

As respondd currently searches its providers in `/lib/gluon/respondd/`, moving the package to the packages repository isn't possible yet. A bit OT: As freifunk-gluon/packages#143 won't be mergable in the near future and thus neither #840, I will prepare a PR for gluon-respondd that moves the provider modules (as proposed in #840) and converts the initscript to procd (to have automatic restarts).

**Comment by lemoer** [comment]
Has anyone tested how stable the results are? Maybe will get the same oscillating issues as in #508 (which are still not fixed yet).

**Comment by corny** [comment]
We monitored these values at a public event with 80 nodes and they are fine.
Values are monotonically increasing unless the channel changes.

**Comment by genofire** [comment]
Hey, here are some nostalgie (collected respondd answers with our [respond-collector](https://github.com/FreifunkBremen/respond-collector) of this event (take a look of the ninth diagram)
screencapture-mgmt-ffhb-de-8080-dashboard-db-breminale-1468785816780

**Comment by jplitza** [comment]
Opened pull request for Gluon-independent package, closing this one.

**Pull request closed without merging**